src/host/utf8ToWideCharParser.cpp

```diff
@@ -1,520 +0,0 @@
-// number of wide chars contained by converted after this function is
-// run, or 0 if an error occurs (or if pBytes is 0).
-// - converted - a valid unique_ptr to store the parsed wide chars
-// in. On error this will contain nullptr instead of an array.
-// Return Value:
-// - <none>
-[[nodiscard]] HRESULT Utf8ToWideCharParser::Parse(_In_reads_(cchBuffer) const byte* const pBytes,
-                                                  _In_ const unsigned int cchBuffer,
-                                                  _Out_ unsigned int& cchConsumed,
-                                                  _Inout_ std::unique_ptr<wchar_t[]>& converted,
-                                                  _Out_ unsigned int& cchConverted)
-{
-    cchConsumed = 0;
-    cchConverted = 0;
-
-    // we can't parse anything if we weren't given any data to parse
-    if (cchBuffer == 0)
-    {
-        return S_OK;
-    }
-    // we shouldn't be parsing if the current codepage isn't UTF8
-    if (_currentCodePage != CP_UTF8)
-    {
-        _currentState = _State::Error;
-    }
-    auto hr = S_OK;
-    try
-    {
-        auto loop = true;
-        unsigned int wideCharCount = 0;
-        _convertedWideChars.reset(nullptr);
-        while (loop)
-        {
-            switch (_currentState)
-            {
-            case _State::Ready:
-                wideCharCount = _ParseFullRange(pBytes, cchBuffer);
-                break;
-            case _State::BeginPartialParse:
-                wideCharCount = _InvolvedParse(pBytes, cchBuffer);
-                break;
-            case _State::Error:
-                hr = E_FAIL;
-                _Reset();
-                wideCharCount = 0;
-                loop = false;
-                break;
-            case _State::Finished:
-                _currentState = _State::Ready;
-                cchConsumed = cchBuffer;
-                loop = false;
-                break;
-            case _State::AwaitingMoreBytes:
-                _currentState = _State::BeginPartialParse;
-                cchConsumed = cchBuffer;
-                loop = false;
-                break;
-            default:
-                _currentState = _State::Error;
-                break;
-            }
-        }
-        converted.swap(_convertedWideChars);
-        cchConverted = wideCharCount;
-    }
-    catch (...)
-    {
-        _Reset();
-        hr = wil::ResultFromCaughtException();
-    }
-    return hr;
-}
-
-// Routine Description:
-// - Determines if ch is a UTF8 lead byte. See _Utf8SequenceSize() for a
-// description of how a lead byte is specified.
-// Arguments:
-// - ch - The byte to test.
-// Return Value:
-// - True if ch is a lead byte, false otherwise.
-bool Utf8ToWideCharParser::_IsLeadByte(_In_ byte ch)
-{
-    auto sequenceSize = _Utf8SequenceSize(ch);
-    return !_IsContinuationByte(ch) &&
-           !_IsAsciiByte(ch) &&
-           sequenceSize > 1 &&
-           sequenceSize <= _UTF8_BYTE_SEQUENCE_MAX;
-}

```

Comment:
I came across this because I was searching for UTF16 "lead bytes" in the project (I forgot the `til::unicodeSomethingSomething` function name for it) and reading this code made me wonder why we still got this.

BTW `_Utf8SequenceSize` is not a trivial function - it's `O(n)`. If I were to inline it into this (and all the other functions) it would look like this:
```cpp
bool Utf8ToWideCharParser::_IsLeadByte(char ch)
{
    char ch2 = ch;
    unsigned int sequenceSize = 0;
    while ((ch2 & 0x80) == 0x80)
    {
        ++sequenceSize;
        ch2 <<= 1;
    }

    return (ch & 0xC0) != 0x80 &&
           (ch & 0x80) == 0x80 &&
           sequenceSize > 1 &&
           sequenceSize <= 4;
}
```

That's certainly one of the `_IsLeadByte` implementations of all time!

Here's how I'd do it - Option 1:
```cpp
bool Utf8ToWideCharParser::_IsLeadByte(char ch)
{
    // 1:1 straight up copy from https://en.wikipedia.org/wiki/UTF-8#Encoding
    return (ch & 0b11100000) == 0b11000000 ||
           (ch & 0b11110000) == 0b11100000 ||
           (ch & 0b11111000) == 0b11110000;
}
```
Option 2:
```cpp
bool Utf8ToWideCharParser::_IsLeadByte(char ch)
{
    // Liberally adopted from https://nullprogram.com/blog/2017/10/06/
    static constexpr bool leads[] = {
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0,
    };
    return leads[ch >> 3];
}
```

Comparison: https://quick-bench.com/q/O02-7dhVB4vYi0--Di79bUikKSA